docs/proposals/volume-provisioning.md

```diff
@@ -121,30 +124,125 @@ We propose that:
     periodically retries finding a matching volume or storage class again until
     a match is found. The claim is `Pending` during this period.
 
-4.  With StorageClass instance, the controller finds volume plugin specified by
-    StorageClass.ProvisionerType.
+4.  With StorageClass instance, the controller updates the claim:
+       * `claim.Spec.Class = storageClass.Name`
+       * `claim.Annotations["volume.beta.kubernetes.io/storage-provisioner"] = storageClass.Provisioner`
+
+* **In-tree provisioning**
+
+   The controller tries to find an internal volume plugin referenced by
+   `storageClass.Provisioner`. If it is found:
+
+  5.  The internal provisioner implements interface`ProvisionableVolumePlugin`,
+      which has a method called `NewProvisioner` that returns a new provisioner.
+
+  6.  The controller calls volume plugin `Provision` with ProvisionerParameters
+      from the `StorageClass` configuration object.
+
+  7.  If `Provision` returns an error, the controller generates an event on the
+      claim and goes back to step 1., i.e. it will retry provisioning
+      periodically.
+
+  8.  If `Provision` returns no error, the controller creates the returned
+      `api.PersistentVolume`, fills its `Class` attribute with `claim.Spec.Class`
+      and makes it already bound to the claim
+
+    1.  If the create operation for the `api.PersistentVolume` fails, it is
+        retried
+
+    2.  If the create operation does not succeed in reasonable time, the
+        controller attempts to delete the provisioned volume and creates an event
+        on the claim
+
+Existing behavior is un-changed for claims that do not specify
+`claim.Spec.Class`.
+
+* **Out of tree provisioning**
+
+  Following step 4. above, the controller tries to find internal plugin for the
+  `StorageClass`. If it is not found, it does not do anything, it just
+  periodically goes to step 1., i.e. tries to find available matching PV.
+
+  External provisioner must have these features:
+
+  * It must have a distinct name, following Kubernetenes plugin naming scheme
+    `<vendor name>/<provisioner name>`, e.g. `kubernetes.io/aws-ebs` or
+    `gluster.org/gluster`.
+
+  * Every time the provisioner needs to report an error to user, it does so by
+    sending an event to the claim that led to the error. It is the same way
+    as internal provisioners do and user gets the same experience.
+
+  * The provisioner must implement also a deleter. It must be able to delete
+    storage assets it created.
```

Comment:
as proposed below, the same plugin / external provisioner would be used to delete volumes.

We could add a separate annotation that provisioner could use to select another deleter, but do we need it?